src/main/java/frc/robot/subsystems/shooter/Shooter.java

```diff
@@ -0,0 +1,134 @@
+        Vector<N1> B = VecBuilder.fill(G * Kt / (OMEGA * J));
+        LinearSystem<N1, N1, N1> stateSpace = new LinearSystem<>(A, B, Matrix.eye(Nat.N1()), new Matrix<>(Nat.N1(), Nat.N1()));
+        KalmanFilter<N1, N1, N1> kalman = new KalmanFilter<>(Nat.N1(), Nat.N1(), stateSpace,
+                VecBuilder.fill(MODEL_TOLERANCE),
+                VecBuilder.fill(ENCODER_TOLERANCE),
+                Constants.ROBOT_TIMEOUT
+        );
+        LinearQuadraticRegulator<N1, N1, N1> lqr = new LinearQuadraticRegulator<>(stateSpace, VecBuilder.fill(VELOCITY_TOLERANCE),
+                VecBuilder.fill(Constants.VOLTAGE),
+                Constants.ROBOT_TIMEOUT // time between loops, DON'T CHANGE
+        );
+        this.stateSpacePredictor = new LinearSystemLoop<>(stateSpace, lqr, kalman, Constants.VOLTAGE, Constants.ROBOT_TIMEOUT);
+    }
+
+    /**
+     * @return the velocity of the shooter. [rps]
+     * @see #setVelocity(double)
+     */
+    public double getVelocity() {
+        return rpsUnitModel.toVelocity(main.getSelectedSensorVelocity());
+    }
+
+    /**
+     * Set the velocity to apply by the motor.
+     *
+     * @param velocity the desired velocity at which the motor will rotate. [rps]
+     * @see #setPower(double)
+     */
+    public void setVelocity(double velocity) {
+        stateSpacePredictor.setNextR(VecBuilder.fill(velocity)); //r = reference
+        stateSpacePredictor.correct(VecBuilder.fill(getVelocity()));
+        stateSpacePredictor.predict(Constants.ROBOT_TIMEOUT); //every 20 ms
+
+        double voltageToApply = stateSpacePredictor.getU(0); // u = input, calculated by the input.
+        // returns the voltage to apply (between 0 and 12)
+        setPower(voltageToApply / Constants.VOLTAGE); // map to be between 0 and 1
+    }
+
+    /**
+     * Set the power to apply by the motor.
+     *
+     * @param power the power at which the motor will rotate. [percentage, between 0 and 1]
+     * @see #stop()
+     */
+    public void setPower(double power) {
+        main.set(TalonFXControlMode.PercentOutput, power, DemandType.ArbitraryFeedForward, Constants.Shooter.ARBITRARY_FEED_FORWARD);
+    }
+
+    /**
+     * Estimate the velocity that the shooter should apply in order to reach the target.
+     * The function created by running an experiment of different velocities for the "most used" distances.
+     * See the experiment results in (TODO: Add the path).
+     *
+     * @param distance the distance from the target.
+     * @return the velocity that should be applied by the shooter in order to reach the target.[rps]
+     */
+    public double estimateVelocity(double distance) {
+        return distance;
+    }
+
+    /**
+     * Get whether the flywheel has reached the desired velocity in order to reach the target.
+     * Also, this function checks whether the flywheel has enough velocity in order to move the motor in first place.
```

Comment:
The current velocity will be non-zero if and only if the shooter moves.
The voltage and current might be non-zero when the shooter doesn't move, but not the velocity.

Reply:
?

Reply:
I'm not sure what you didn't understand.
If it's about "if and only if", then it is "biconditional logical connective between statements, where either both statements are true or both are false" ([source](https://en.wikipedia.org/wiki/If_and_only_if)).

I was talking about this condition: `currentVelocity > Constants.Shooter.MINIMAL_VELOCITY`.
What I meant is that if the flywheel moves, its velocity is greater than zero, so this condition is redundant.